### ISSUE TITLE
Response parsing error - no matching QueryResponseKind 

### ISSUE THREAD

**Issue by blackghost1987** [near/near-jsonrpc-client-rs#74]
I'm trying to send a test RpcQueryRequest to the meme museum demo contract at `museum.testnet`. I'm trying to query the `get_meme_list` function, but I get an unhelpful error message, without any details on what am I doing wrong:

`TransportError(RecvError(ResponseParseError(ResultParseError(Error("data did not match any variant of untagged enum QueryResponseKind", line: 0, column: 0)))`

Transport should be okay, because when I use a RpcTransactionStatusRequest I get a parsed response.

I think my call should be the equivalent of calling:
`near view museum.testnet get_meme_list`
So I tried to call that, and I got the following error:
`wasm execution failed with error: FunctionCallError(HostError(GasLimitExceeded))`

Maybe this is the error which is not handled by the client? But how could there be a gas limit error on a query? Queries should not cost any gas, right?

So I assume that this unmatched QueryResponseKind error should not happen normally and the client should be able to decode all possible responses from the RPC node. What steps could I do to debug this issue? Is there like detailed logging or any way to print the raw response to see what kind of QueryResponseKind is tripping up the parser?

My dependencies are the following:
```
near-jsonrpc-client = "0.3"
near-jsonrpc-primitives = "0.12"
near-primitives = "0.12"
```

**Comment by blackghost1987** [comment]
I've checked to code and there seems to be no logging at all, so I had to monkey-patch it and print out the raw response that way. This is what I got:
```
Response {
    jsonrpc: Version,
    result: Ok(
        Object({
            "block_hash": String(
                "4jKqW75Mq5urCDhNL3HdonNvVFxvVdgZcgjsSiSfrhjY",
            ),
            "block_height": Number(
                86981756,
            ),
            "error": String(
                "wasm execution failed with error: FunctionCallError(HostError(GasLimitExceeded))",
            ),
            "logs": Array([]),
        }),
    ),
    id: String(
        "471a4bec-000a-42e4-803a-125741ac10bf",
    ),
}
```

Looks like the error handling is messed up, it tries to parse the error as a successful RpcQueryResponse for some reason. The Object does have a `block_hash` and a `block_height`, but it has an `error` instead of the `kind` field, that's why the parser is complaining. I'm not sure this should be an `Ok` result at all.

Is this a bug or a missing feature?

**Comment by miraclx** [comment]
Hi @blackghost1987..

Yes, this is a recognized issue. Here's some context: https://github.com/near/nearcore/issues/6384.

Basically, while nearcore JSON RPC returns structured errors for all other cases, in the case of `RpcQueryError`, for two fields in particular - [`ContractExecutionError`](https://github.com/near/nearcore/blob/78c4b0c5938f7a8af37927e554207d68e48ffafa/chain/jsonrpc/src/lib.rs#L143-L152) and [`UnknownAccessKey`](https://github.com/near/nearcore/blob/78c4b0c5938f7a8af37927e554207d68e48ffafa/chain/jsonrpc/src/lib.rs#L153-L165), it returns a makeshift JSON response as a `result`, with an `error` embedded within.

Whereas, the client expects appropriately structured data that's deserializable into either the `RpcQueryResponse` or `RpcQueryError` structs.

A fix could potentially mean a breaking change on the nearcore JSON RPC side, in order to have higher quality, structured errors on the client side (instead of just plain strings).

But it will be delivered soon. In the meantime, your patch highlights the issue you're having – `FunctionCallError(HostError(GasLimitExceeded))`.

You should be unblocked now, sorry for the inconvenience.

**Comment by blackghost1987** [comment]
Yeah, I understand the response parsing problem now, thanks!

I still don't understand the gas issue on a query, I thought queries should be free. Is there a maximum gas value which limits the complexity of queries or something? I couldn't find anything mentioning this in the docs, could you point me in the right direction?

I know this is not really related to the library, it's more like a general Near question, so of course I'm closing the issue anyway.

**Comment by frol** [comment]
@blackghost1987 https://stackoverflow.com/questions/68017713/can-i-run-a-near-node-with-a-higher-gas-limit-for-view-calls

**Comment by miraclx** [comment]
@blackghost1987, fixed on #82.
